gnu/man/uniq.man
NAME
     uniq - remove duplicate lines from a sorted file

SYNOPSIS
     uniq [-cdu] [-f skip-fields] [-s skip-chars]  [-w  check-chars]  [-#skip-
     fields]   [+#skip-chars]   [--count]   [--repeated]  [--unique]  [--skip-
     fields=skip-fields] [--skip-chars=skip-chars] [--check-chars=check-chars]
     [infile] [outfile]

DESCRIPTION
     This manual page documents the GNU version  of  uniq.   uniq  prints  the
     unique  lines  in  a  sorted  file,  discarding  all  but one of a run of
     matching lines.  It can optionally show only lines  that  appear  exactly
     once,  or  lines  that appear more than once.  uniq requires sorted input
     because it compares only consecutive lines.

     If the output file is not specified, uniq writes to the standard  output.
     If the input file is not specified, it reads from the standard input.

     OPTIONS

     -u, --unique
          Only print unique lines.

     -d, --repeated
          Only print duplicate lines.

     -c, --count
          Print the number of times each line occurred along with the line.

     -, -f, --skip-fields=number
          In this option, number is an  integer  representing  the  number  of
          fields  to  skip  over  before  checking  for uniqueness.  The first
          number fields, along with any blanks found before number  fields  is
          reached,  are skipped over and not counted.  Fields are defined as a
          strings of non-space, non-tab characters, that  are  separated  from
          each other by spaces and tabs.

     --, -s, +skip-chars=number
          In this option,  number  is  an  integer  represent  the  number  of
          characters  to  skip over before checking for uniqueness.  The first
          number  characters,  along  with  any  blanks  found  before  number
          characters is reached, are skipped over and not counted.  If you use
          both the field and character skipping options,  fields  are  skipped
          over first.

     -w, +check-chars=number
          Specify the number of characters to  compare  in  the  lines,  after
          skipping  any  specified fields and characters.  Normally the entire
          rest of the lines are compared.

     The long-named options can be introduced with `+' as well  as  `--',  for
     compatibility with previous releases.  Eventually support for `+' will be
     removed, because it is incompatible with the POSIX.2 standard.


gnu/ChangeLog
Sun Feb  2 21:22:01 1992  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* Version 1.2.

	* nl.c: Support multiple files and "-" for stdin.
	(main): Check for read and write errors.
	(nl_file): New function.

Wed Jan 29 10:09:10 1992  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tr.c (main): -t option was called -b in getopt spec.
	(validate): Don't warn that set1 is longer than set2.

	* tr.c: Rename --sysv-string2-truncate to --truncate-string1.

Fri Jan 17 16:29:05 1992  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* nl.c: New program from bin-src.

	* nl.c (main): Use a struct linebuffer for line_buf.
	(process_file): Use readline instead of fgets, to preserve NULs.
	(check_section): Use memcmp instead of strncmp.
	(proc_text): Print line_buf with fwrite instead of printf.

	* nl.c (main): Usage message if too many args given.  Check
	for error in closing input file.  Lengths of section delimiter
	strings were 1 too large.  Take separator_str into account in
	length of print_no_line_fmt.
	(build_print_fmt): Allocate space for print_fmt, in case
	separator_str is long.
	(proc_text): A blank line is one that contains nothing, not
	even nonprinting characters.

Fri Jan 17 01:04:22 1992  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* All programs: Document `--' instead of `+' to introduce
	long-named options, in usage messages.

	* sum.c (bsd_sum_file): Renamed from sum_file.
	(sysv_sum_file): New function.
	(main): Recognize an option to select between the algorithms.

Sun Jan  5 17:41:18 1992  Jim Meyering (meyering at churchy.gnu.ai.mit.edu)

	* pr.c (close_file, print_page): Fixed bug that had the command
	yes |head |pr -t printing "yyyyyyyyyy".
	* (print_page): Fixed bug that had pr -3 -a printing two too few
	trailer lines per page.
	* (main): Added restriction that -a and -m are incompatible.
	Although the POSIX spec doesn't explicitly say they shouldn't
	be used together, it says -a modifies the -column option and
	that -column shouldn't be used with -m.

Thu Jan  2 15:23:59 1992  David J. MacKenzie  (djm at albert.gnu.ai.mit.edu)

	* nl.c: Include regex.h after, not before, sys/types.h.

Thu Jan  2 12:18:10 1992  Tom Lord  (lord at geech.gnu.ai.mit.edu)

	* sort.c (fillbuf) return bytes buffered instead of bytes read.

Fri Dec 27 22:53:36 1991  Jim Kingdon  (kingdon at geech.gnu.ai.mit.edu)

	* sort.c (LINEALLOC): New #define.
	(struct lines): New field ``limit''.
	(initlines): Set it from new arg ``limit''.
	(sort, mergefps, checkfp): Pass new arg to initlines().
	(findlines): Don't realloc past lines->limit.

Tue Dec 24 01:24:03 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tac.c, sort.c, csplit.c, system.h: Change POSIX ifdefs to
	HAVE_UNISTD_H and _POSIX_VERSION.

	* xwrite.c: Change POSIX ifdef to HAVE_UNISTD_H.

Sat 14 Dec 1991 11:46:42  Jim Meyering  (meyering at wombat)

	* tr.c: Fixed an inaccurate comment on posix_pedantic.

Thu 12 Dec 1991 21:15:20  Jim Meyering  (meyering at hal)

	* tr.c: Changed underscores to hyphens in long option name
	"sysv_string2_truncate".

Wed Dec 11 13:33:34 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tac.c (main): Set obscure_syntax to tell re_search to
	allocate memory for the group registers.

Fri Dec  6 18:26:27 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tac.c, sort.c, csplit.c [POSIX]:  Use sigaction instead of
	signal, which POSIX doesn't have.
	* sort.c: Replace inthandler and huphandler with sighandler.
	* csplit.c (main): Only handle signals if they were not being
	ignored.

	* tr.c: POSIX_ME_HARDER -> POSIXLY_CORRECT.

Wed Dec  4 00:47:47 1991  Jim Meyering  (meyering at wombat)

	* tr.c (unquote): Reformat code so it doesn't go beyond column 80.
	* tr.c (squeeze_filter): Comment a little on why it's better
	to step through the input by two.
	* tr.c (set_initialize): Write a comment describing the function.
	* tr.c: Eliminated the variable `portability_warnings' and replaced
	references to it by references to `!posix_pedantic'.  One of the
	uses of portability_warnings had been wrong.

Tue Dec  3 14:03:35 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tr.c: New program.

Sun Dec  1 15:07:35 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* linebuffer.[ch] (freebuffer): New function (used by cron).

Thu Oct 17 22:30:22 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* system.h, configure, Makefile.in: Don't include memory.h if
	STDC_HEADERS, removing need for MEMORY_H_MISSING.

Thu 17 Oct 1991 16:53:07  Jim Meyering  (meyering at wombat)

	* pr.c (print_page): REALLY fixed `extra newline at EOF' problem.
	Somehow, part of my patch didn't make it last time.

Sat Oct 12 12:04:47 1991  David J. MacKenzie  (djm at churchy.gnu.ai.mit.edu)

	* tail.c (pipe_lines, pipe_bytes): Initialize `first->next'.

	* cmp.c (cmp): Print EOF message to stderr, not stdout, for
	POSIX 1003.2.11.2.

	* sort.c (xfwrite): fwrite never returns < 0, so check if
	number written is number we asked to write.
	(fillbuf, main): fread never returns < 0, so check ferror instead.
	From Rainer Orth.

Tue Oct  8 18:07:08 1991  Jim Meyering  (meyering at churchy)

	* pr.c (print_page): Really fixed `extra newline at EOF' problem.
	* (store_columns): Fixed bug that caused `pr -b -2' to coredump
	on files of certain lengths.

Fri Oct  4 22:30:25 1991  Jim Meyering  (meyering at churchy)

	* pr.c (print_page): Fixed to not add single spurious newline
	at EOF when using -t.

Wed Oct  2 01:02:05 1991  David J. MacKenzie  (djm at apple-gunkies)

	* pr.c (print_page): Don't pad the page if -t given.

	* csplit.c (load_buffer), sort.c (mergefps): Use bcopy, not memcpy.

	* memchr.c (memchr): Fix typo.

Thu Sep 26 12:35:00 1991  David J. MacKenzie  (djm at churchy.gnu.ai.mit.edu)


gnu/src/uniq.c
/* uniq -- remove duplicate lines from a sorted file
   Copyright (C) 1986, 1991 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.  */

/* Written by Richard Stallman and David MacKenzie. */

#define _GNU_SOURCE
#include <ctype.h>
#ifndef isblank
#define isblank(c) ((c) == ' ' || (c) == '\t')
#endif
#include <stdio.h>
#include <getopt.h>
#include <sys/types.h>
#include "system.h"
#include "linebuffer.h"

#ifndef min
#define min(x, y) ((x) < (y) ? (x) : (y))
#endif

char *find_field ();
int different ();
void check_file ();
void error ();
void usage ();
void writeline ();

/* Number of fields to skip on each line when doing comparisons. */
int skip_fields;

/* Number of chars to skip after skipping any fields. */
int skip_chars;

/* Number of chars to compare; if 0, compare the whole lines. */
int check_chars;

enum countmode
{
  count_occurrences,		/* -c Print count before output lines. */
  count_none			/* Default.  Do not print counts. */
};

/* Whether and how to precede the output lines with a count of the number of
   times they occurred in the input. */
enum countmode countmode;

enum output_mode
{
  output_repeated,		/* -d Only lines that are repeated. */
  output_unique,		/* -u Only lines that are not repeated. */
  output_all			/* Default.  Print first copy of each line. */
};

/* Which lines to output. */
enum output_mode mode;

/* The name this program was run with. */
char *program_name;

struct option longopts[] =
{
  {"count", 0, NULL, 'c'},
  {"repeated", 0, NULL, 'd'},
  {"unique", 0, NULL, 'u'},
  {"skip-fields", 1, NULL, 'f'},
  {"skip-chars", 1, NULL, 's'},
  {"check-chars", 1, NULL, 'w'},
  {NULL, 0, NULL, 0}
};

void
main (argc, argv)
     int argc;
     char *argv[];
{
  int optc;
  char *infile = "-", *outfile = "-";

  program_name = argv[0];
  skip_chars = 0;
  skip_fields = 0;
  check_chars = 0;
  mode = output_all;
  countmode = count_none;

  while ((optc = getopt_long (argc, argv, "0123456789cdf:s:uw:", longopts,
			      (int *) 0)) != EOF)
    {
      switch (optc)
	{
	case '0':
	case '1':
	case '2':
	case '3':
	case '4':
	case '5':
	case '6':
	case '7':
	case '8':
	case '9':
	  skip_fields = skip_fields * 10 + optc - '0';
	  break;

	case 'c':
	  countmode = count_occurrences;
	  break;

	case 'd':
	  mode = output_repeated;
	  break;

	case 'f':		/* Like '-#'. */
	  skip_fields = atoi (optarg);
	  break;

	case 's':		/* Like '+#'. */
	  skip_chars = atoi (optarg);
	  break;

	case 'u':
	  mode = output_unique;
	  break;

	case 'w':
	  check_chars = atoi (optarg);
	  break;

	default:
	  usage ();
	}
    }

  while (optind < argc && argv[optind][0] == '+')
    skip_chars = atoi (argv[optind++]);

  if (optind == argc && isatty(0))
    usage ();			/* No arguments. */

  if (optind < argc)
    infile = argv[optind++];

  if (optind < argc)
    outfile = argv[optind++];

  if (optind < argc)
    usage ();			/* Extra arguments. */

  check_file (infile, outfile);

  exit (0);
}

/* Process input file INFILE with output to OUTFILE.
   If either is "-", use the standard I/O stream for it instead. */

void
check_file (infile, outfile)
     char *infile, *outfile;
{
  FILE *istream;
  FILE *ostream;
  struct linebuffer lb1, lb2;
  struct linebuffer *thisline, *prevline, *exch;
  char *prevfield, *thisfield;
  int prevlen, thislen;
  int match_count = 0;

  if (!strcmp (infile, "-"))
    istream = stdin;
  else
    istream = fopen (infile, "r");
  if (istream == NULL)
    error (1, errno, "%s", infile);

  if (!strcmp (outfile, "-"))
    ostream = stdout;
  else
    ostream = fopen (outfile, "w");
  if (ostream == NULL)
    error (1, errno, "%s", outfile);

  thisline = &lb1;
  prevline = &lb2;

  initbuffer (thisline);
  initbuffer (prevline);

  if (readline (prevline, istream) == 0)
    goto closefiles;
  prevfield = find_field (prevline);
  prevlen = prevline->length - (prevfield - prevline->buffer);

  while (!feof (istream))
    {
      if (readline (thisline, istream) == 0)
	break;
      thisfield = find_field (thisline);
      thislen = thisline->length - (thisfield - thisline->buffer);
      if (!different (thisfield, prevfield, thislen, prevlen))
	match_count++;
      else
	{
	  writeline (prevline, ostream, match_count);
	  match_count = 0;

	  exch = prevline;
	  prevline = thisline;
	  thisline = exch;
	  prevfield = thisfield;
	  prevlen = thislen;
	}
    }

  writeline (prevline, ostream, match_count);

 closefiles:
  if (ferror (istream) || fclose (istream) == EOF)
    error (1, errno, "error reading %s", infile);

  if (ferror (ostream) || fclose (ostream) == EOF)
    error (1, errno, "error writing %s", outfile);

  free (lb1.buffer);
  free (lb2.buffer);
}

/* Given a linebuffer LINE,
   return a pointer to the beginning of the line's field to be compared. */

char *
find_field (line)
     struct linebuffer *line;
{
  register int count;
  register char *lp = line->buffer;
  register int size = line->length;
  register int i = 0;

  for (count = 0; count < skip_fields && i < size; count++)
    {
      while (i < size && isblank (lp[i]))
	i++;
      while (i < size && !isblank (lp[i]))
	i++;
    }

  for (count = 0; count < skip_chars && i < size; count++)
    i++;

  return lp + i;
}

/* Return zero if two strings OLD and NEW match, nonzero if not.
   OLD and NEW point not to the beginnings of the lines
   but rather to the beginnings of the fields to compare.
   OLDLEN and NEWLEN are their lengths. */

int
different (old, new, oldlen, newlen)
     char *old;
     char *new;
     int oldlen;
     int newlen;
{
  register int order;

  if (check_chars)
    {
      if (oldlen > check_chars)
	oldlen = check_chars;
      if (newlen > check_chars)
	newlen = check_chars;
    }
  order = memcmp (old, new, min (oldlen, newlen));
  if (order == 0)
    return oldlen - newlen;
  return order;
}

/* Output the line in linebuffer LINE to stream STREAM
   provided that the switches say it should be output.
   If requested, print the number of times it occurred, as well;
   LINECOUNT + 1 is the number of times that the line occurred. */

void
writeline (line, stream, linecount)
     struct linebuffer *line;
     FILE *stream;
     int linecount;
{
  if ((mode == output_unique && linecount != 0)
      || (mode == output_repeated && linecount == 0))
    return;

  if (countmode == count_occurrences)
    fprintf (stream, "%7d\t", linecount + 1);

  fwrite (line->buffer, sizeof (char), (size_t) line->length, stream);
  putc ('\n', stream);
}

void
usage ()
{
  print_version();
  fprintf (stderr, "\
Usage: %s [-cdu] [-f skip-fields] [-s skip-chars] [-w check-chars]\n\
       [-#skip-fields] [+#skip-chars] [--count] [--repeated] [--unique]\n\
       [--skip-fields=skip-fields] [--skip-chars=skip-chars]\n\
       [--check-chars=check-chars] [infile] [outfile]\n",
	   program_name);
  exit (1);
}
